zerver/lib/mention.py

```diff
@@ -21,6 +21,22 @@ class FullNameInfo:
     full_name: str
 
 
+@dataclass
+class UserFilter:
+    id: Optional[int]
+    full_name: Optional[str]
+
+    def Q(self) -> Any:
+        if self.full_name is not None and self.id is not None:
+            return Q(full_name__iexact=self.full_name, id=self.id)
+        elif self.id is not None:
+            return Q(id=self.id)
+        elif self.full_name is not None:
+            return Q(full_name__iexact=self.full_name)
+        else:
+            raise AssertionError("totally empty filter makes no sense")
```

Comment:
I don't understand where you're going with this extraction; it seems like having the `Q` object return an `Any` isn't ideal.

Reply:
I can explain this better over chat, but the idea behind this extraction is to allow us to efficiently send a bunch of PMs to a bunch of users when we subscribe a bunch of users to a stream.

It sucks that the Q() function returns an Any type, but mypy won't accept `object`, and Django and mypy generally don't play nice.  (I can explain that in person or over chat.)

Reply:
In the next day or two I will push the rest of the PR, and this decision will make more sense.  In the meantime, the first few commits of this PR are probably no-brainers.

Reply:
The right annotation is `def Q(self) -> Q`, since the value has type `django.db.models.Q` at runtime.

For now, mypy will treat that as equivalent to `-> Any` because it doesn’t have stubs for `django.db.models.Q`, but it will be its own type once we add django-stubs (#19552).

Reply:
> The right annotation is `def Q(self) -> Q`, since the value has type `django.db.models.Q` at runtime.
> 
> For now, mypy will treat that as equivalent to `-> Any` because it doesn’t have stubs for `django.db.models.Q`, but it will be its own type once we add django-stubs (#18777).

Arrrrrrrrrrrrrrrrrg! I knew that Q was probably Any at heart but it somehow didn’t occur to me that the type of Q is Q.  Omg

Reply:
Who is buried in Grant’s Tomb?

Reply:
I always wondered what was happening under the hood.  See `/django/db/models/query_utils.py`.

Ok, next push will use Q as the return type.  (For some reason I was thinking Q was a function returning some other data type.)

~~~ py
 57 class Q(tree.Node):
 58     """
 59     Encapsulate filters as objects that can then be combined logically (using
 60     `&` and `|`).
 61     """
 62     # Connection types
 63     AND = 'AND'
 64     OR = 'OR'
 65     default = AND
 66     conditional = True
 67
 68     def __init__(self, *args, _connector=None, _negated=False, **kwargs):
 69         super().__init__(children=[*args, *sorted(kwargs.items())], connector=_connector, negated=_negated)
 70
[...]